File: internal/core/memory_consolidation.go
package core

// memory_consolidation.go — ConsolidationPipeline, MemoryIndex, DecayManager,
// ImportanceScorer.
//
// Go owns scheduling and lifecycle.  The actual BOCPD changepoint math stays in
// Python and is called via the ChangePointDetector bridge interface.

import (
	"log/slog"
	"math"
	"sync"
	"time"

	"github.com/google/uuid"
)

// ---------------------------------------------------------------------------
// Bridge interface — Python implements BOCPD on the other side
// ---------------------------------------------------------------------------

// ChangePointDetector is called by the consolidation pipeline to detect regime
// shifts in a stream of importance values.  Python implements this via
// Connect-RPC; in tests a stub can be injected.
type ChangePointDetector interface {
	// DetectChangePoints returns the indices in values where changepoints were
	// detected.  Returns nil if none detected or if the detector is unavailable.
	DetectChangePoints(values []float64) ([]int, error)
}

// ---------------------------------------------------------------------------
// MemoryRecord
// ---------------------------------------------------------------------------

// MemoryRecord is a single memory entry that may live in short-term or
// long-term storage.
type MemoryRecord struct {
	MemoryID     string
	CreatedAt    time.Time
	LastAccessed time.Time
	Importance   float64 // 0.0–1.0
	RegimeTag    string
	SignalType   string
	Content      map[string]any
	Tags         []string
	IsLandmark   bool  // if true, protected from decay below LandmarkFloor
	AccessCount  int
	Consolidated bool // true once promoted to long-term store
}

// NewMemoryRecord constructs a MemoryRecord with sensible defaults.
func NewMemoryRecord(regimeTag, signalType string, content map[string]any, tags []string, importance float64, isLandmark bool) MemoryRecord {
	now := time.Now()
	return MemoryRecord{
		MemoryID:     uuid.New().String(),
		CreatedAt:    now,
		LastAccessed: now,
		Importance:   clamp(importance, 0, 1),
		RegimeTag:    regimeTag,
		SignalType:   signalType,
		Content:      content,
		Tags:         tags,
		IsLandmark:   isLandmark,
	}
}

// ---------------------------------------------------------------------------
// MemoryIndex — fast in-memory lookup
// ---------------------------------------------------------------------------

const bucketSeconds = 86_400 // one day

// MemoryIndex is an in-memory inverted index for fast lookup by regime tag,
// signal type, or time range.
type MemoryIndex struct {
	mu       sync.RWMutex
	byRegime map[string]map[string]struct{}
	bySignal map[string]map[string]struct{}
	byTime   map[int64]map[string]struct{} // day bucket → IDs
}

// NewMemoryIndex initialises an empty MemoryIndex.
func NewMemoryIndex() *MemoryIndex {
	return &MemoryIndex{
		byRegime: make(map[string]map[string]struct{}),
		bySignal: make(map[string]map[string]struct{}),
		byTime:   make(map[int64]map[string]struct{}),
	}
}

func (idx *MemoryIndex) Add(r MemoryRecord) {
	idx.mu.Lock()
	defer idx.mu.Unlock()
	addToSet(idx.byRegime, r.RegimeTag, r.MemoryID)
	addToSet(idx.bySignal, r.SignalType, r.MemoryID)
	bucket := r.CreatedAt.Unix() / bucketSeconds
	addToSet64(idx.byTime, bucket, r.MemoryID)
}

func (idx *MemoryIndex) Remove(r MemoryRecord) {
	idx.mu.Lock()
	defer idx.mu.Unlock()
	removeFromSet(idx.byRegime, r.RegimeTag, r.MemoryID)
	removeFromSet(idx.bySignal, r.SignalType, r.MemoryID)
	bucket := r.CreatedAt.Unix() / bucketSeconds
	removeFromSet64(idx.byTime, bucket, r.MemoryID)
}

// ByRegime returns a copy of IDs for the given regime tag.
func (idx *MemoryIndex) ByRegime(tag string) map[string]struct{} {
	idx.mu.RLock()
	defer idx.mu.RUnlock()
	return copySet(idx.byRegime[tag])
}

// BySignal returns a copy of IDs for the given signal type.
func (idx *MemoryIndex) BySignal(sigType string) map[string]struct{} {
	idx.mu.RLock()
	defer idx.mu.RUnlock()
	return copySet(idx.bySignal[sigType])
}

// ByTimeRange returns IDs for records created between start and end.
func (idx *MemoryIndex) ByTimeRange(start, end time.Time) map[string]struct{} {
	idx.mu.RLock()
	defer idx.mu.RUnlock()
	startBucket := start.Unix() / bucketSeconds
	endBucket := end.Unix() / bucketSeconds
	out := map[string]struct{}{}
	for b := startBucket; b <= endBucket; b++ {
		for id := range idx.byTime[b] {
			out[id] = struct{}{}
		}
	}
	return out
}

// Intersect returns the intersection of multiple ID sets.
func Intersect(sets ...map[string]struct{}) map[string]struct{} {
	if len(sets) == 0 {
		return nil
	}
	result := copySet(sets[0])
	for _, s := range sets[1:] {
		for id := range result {
			if _, ok := s[id]; !ok {
				delete(result, id)
			}
		}
	}
	return result
}

// IndexStats returns basic stats about the index.
func (idx *MemoryIndex) IndexStats() map[string]any {
	idx.mu.RLock()
	defer idx.mu.RUnlock()
	regimes := make(map[string]int, len(idx.byRegime))
	for k, v := range idx.byRegime {
		regimes[k] = len(v)
	}
	signals := make(map[string]int, len(idx.bySignal))
	for k, v := range idx.bySignal {
		signals[k] = len(v)
	}
	return map[string]any{
		"regimes":      regimes,
		"signal_types": signals,
		"time_buckets": len(idx.byTime),
	}
}

// ---------------------------------------------------------------------------
// DecayManager
// ---------------------------------------------------------------------------

// DecayManager applies exponential importance decay to MemoryRecords.
//
//	I(t) = max(floor, I₀ × exp(−λ × Δt))
//
// where λ = ln(2) / halfLife.
type DecayManager struct {
	decayLambda   float64
	floor         float64
	landmarkFloor float64
}

// NewDecayManager creates a DecayManager with the given configuration.
func NewDecayManager(halfLife time.Duration, floor, landmarkFloor float64) (*DecayManager, error) {
	if halfLife <= 0 {
		return nil, &ValidationError{Field: "halfLife", Message: "must be positive"}
	}
	return &DecayManager{
		decayLambda:   math.Log(2) / halfLife.Seconds(),
		floor:         floor,
		landmarkFloor: landmarkFloor,
	}, nil
}

// Decay mutates r.Importance in place.
func (dm *DecayManager) Decay(r *MemoryRecord, now time.Time) {
	elapsed := math.Max(0, now.Sub(r.CreatedAt).Seconds())
	decayed := r.Importance * math.Exp(-dm.decayLambda*elapsed)
	floor := dm.floor
	if r.IsLandmark {
		floor = dm.landmarkFloor
	}
	r.Importance = math.Max(floor, decayed)
}

// DecayAll applies Decay to every record in the slice.
func (dm *DecayManager) DecayAll(records []*MemoryRecord, now time.Time) {
	for _, r := range records {
		dm.Decay(r, now)
	}
}

// ShouldPrune returns true if the record is below threshold and not a landmark.
func (dm *DecayManager) ShouldPrune(r *MemoryRecord, threshold float64) bool {
	if r.IsLandmark {
		return false
	}
	return r.Importance < threshold
}

// ---------------------------------------------------------------------------
// ImportanceScorer
// ---------------------------------------------------------------------------

// ImportanceScorer ranks MemoryRecords by relevance to the current regime.
//
// score = wImportance×importance + wRecency×recency + wRegime×regimeMatch + wAccess×access
type ImportanceScorer struct {
	wImportance float64
	wRecency    float64
	wRegime     float64
	wAccess     float64
	recencyLambda float64
}

const recencyHalfLife = 30 * 24 * time.Hour

// NewImportanceScorer creates a scorer with the given weights (must sum to 1.0).
func NewImportanceScorer(wImportance, wRecency, wRegime, wAccess float64) (*ImportanceScorer, error) {
	total := wImportance + wRecency + wRegime + wAccess
	if math.Abs(total-1.0) > 1e-6 {
		return nil, &ValidationError{Field: "weights", Message: "must sum to 1.0"}
	}
	return &ImportanceScorer{
		wImportance:   wImportance,
		wRecency:      wRecency,
		wRegime:       wRegime,
		wAccess:       wAccess,
		recencyLambda: math.Log(2) / recencyHalfLife.Seconds(),
	}, nil
}

// DefaultImportanceScorer returns a scorer with canonical weights.
func DefaultImportanceScorer() *ImportanceScorer {
	s, _ := NewImportanceScorer(0.4, 0.3, 0.2, 0.1)
	return s
}

// Score computes the relevance score of a single record.
func (s *ImportanceScorer) Score(r *MemoryRecord, currentRegime string, now time.Time) float64 {
	age := math.Max(0, now.Sub(r.CreatedAt).Seconds())
	recency := math.Exp(-s.recencyLambda * age)
	regimeMatch := 0.0
	if r.RegimeTag == currentRegime {
		regimeMatch = 1.0
	}
	access := math.Tanh(float64(r.AccessCount) / 10.0)
	return s.wImportance*r.Importance + s.wRecency*recency + s.wRegime*regimeMatch + s.wAccess*access
}

// ScoredRecord pairs a score with a MemoryRecord pointer.
type ScoredRecord struct {
	Score  float64
	Record *MemoryRecord
}

// Rank returns records sorted descending by score, optionally capped at topK.
func (s *ImportanceScorer) Rank(records []*MemoryRecord, currentRegime string, now time.Time, topK int) []ScoredRecord {
	out := make([]ScoredRecord, 0, len(records))
	for _, r := range records {
		out = append(out, ScoredRecord{Score: s.Score(r, currentRegime, now), Record: r})
	}
	// sort descending by score (insertion sort; N is typically small)
	for i := 1; i < len(out); i++ {
		for j := i; j > 0 && out[j].Score > out[j-1].Score; j-- {
			out[j], out[j-1] = out[j-1], out[j]
		}
	}
	if topK > 0 && topK < len(out) {
		out = out[:topK]
	}
	return out
}

// ---------------------------------------------------------------------------
// ConsolidationPipeline
// ---------------------------------------------------------------------------

// ConsolidationResult summarises a single consolidation run.
type ConsolidationResult struct {
	RunID            string
	Timestamp        time.Time
	ShortTermBefore  int
	ShortTermAfter   int
	LongTermBefore   int
	LongTermAfter    int
	Pruned           int
	Consolidated     int
}

// ConsolidationConfig holds configuration for the pipeline.
type ConsolidationConfig struct {
	MinAge             time.Duration // memory must be this old before promotion
	PromotionThreshold float64       // minimum importance to promote
	PruneThreshold     float64       // prune short-term below this
	Interval           time.Duration // minimum time between consolidation runs
	DecayHalfLife      time.Duration
	LandmarkFloor      float64
}

// DefaultConsolidationConfig returns sensible defaults.
func DefaultConsolidationConfig() ConsolidationConfig {
	return ConsolidationConfig{
		MinAge:             time.Hour,
		PromotionThreshold: 0.3,
		PruneThreshold:     0.05,
		Interval:           15 * time.Minute,
		DecayHalfLife:      7 * 24 * time.Hour,
		LandmarkFloor:      0.3,
	}
}

// ConsolidationPipeline periodically merges short-term memories into
// long-term storage.  Go owns all scheduling and lifecycle; Python BOCPD
// math is called via the ChangePointDetector interface.
type ConsolidationPipeline struct {
	cfg     ConsolidationConfig
	decay   *DecayManager
	scorer  *ImportanceScorer
	index   *MemoryIndex
	cpd     ChangePointDetector // may be nil — disables changepoint enrichment

	mu               sync.Mutex
	shortTerm        map[string]*MemoryRecord
	longTerm         map[string]*MemoryRecord
	lastConsolidation time.Time
	history          []ConsolidationResult
}

// NewConsolidationPipeline creates a pipeline with the given config.
// cpd may be nil; if nil, changepoint detection is skipped.
func NewConsolidationPipeline(cfg ConsolidationConfig, cpd ChangePointDetector) (*ConsolidationPipeline, error) {
	dm, err := NewDecayManager(cfg.DecayHalfLife, 0, cfg.LandmarkFloor)
	if err != nil {
		return nil, err
	}
	return &ConsolidationPipeline{
		cfg:               cfg,
		decay:             dm,
		scorer:            DefaultImportanceScorer(),
		index:             NewMemoryIndex(),
		cpd:               cpd,
		shortTerm:         make(map[string]*MemoryRecord),
		longTerm:          make(map[string]*MemoryRecord),
		lastConsolidation: time.Now(),
	}, nil
}

// Add ingests a new memory into short-term storage.
func (p *ConsolidationPipeline) Add(r MemoryRecord) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.shortTerm[r.MemoryID] = &r
	p.index.Add(r)
	slog.Debug("memory added", "id", r.MemoryID[:8], "regime", r.RegimeTag, "signal", r.SignalType, "importance", r.Importance)
}

// Get retrieves a record from either store; updates access tracking.
func (p *ConsolidationPipeline) Get(id string) (*MemoryRecord, bool) {
	p.mu.Lock()
	defer p.mu.Unlock()
	r := p.shortTerm[id]
	if r == nil {
		r = p.longTerm[id]
	}
	if r != nil {
		r.LastAccessed = time.Now()
		r.AccessCount++
	}
	return r, r != nil
}

// Query returns top-k records matching the given criteria, ranked by score.
func (p *ConsolidationPipeline) Query(regimeTag, signalType string, start, end *time.Time, currentRegime string, topK int) []ScoredRecord {
	p.mu.Lock()
	defer p.mu.Unlock()

	var candidateIDs map[string]struct{}

	intersect := func(ids map[string]struct{}) {
		if candidateIDs == nil {
			candidateIDs = ids
		} else {
			candidateIDs = Intersect(candidateIDs, ids)
		}
	}

	if regimeTag != "" {
		intersect(p.index.ByRegime(regimeTag))
	}
	if signalType != "" {
		intersect(p.index.BySignal(signalType))
	}
	if start != nil && end != nil {
		intersect(p.index.ByTimeRange(*start, *end))
	}

	now := time.Now()
	var records []*MemoryRecord
	if candidateIDs == nil {
		for _, r := range p.shortTerm {
			records = append(records, r)
		}
		for _, r := range p.longTerm {
			records = append(records, r)
		}
	} else {
		for id := range candidateIDs {
			if r, ok := p.shortTerm[id]; ok {
				records = append(records, r)
			} else if r, ok := p.longTerm[id]; ok {
				records = append(records, r)
			}
		}
	}

	return p.scorer.Rank(records, currentRegime, now, topK)
}

// MaybeConsolidate runs consolidation if the interval has elapsed.
func (p *ConsolidationPipeline) MaybeConsolidate() (*ConsolidationResult, error) {
	p.mu.Lock()
	elapsed := time.Since(p.lastConsolidation)
	p.mu.Unlock()
	if elapsed < p.cfg.Interval {
		return nil, nil
	}
	result, err := p.Consolidate()
	return &result, err
}

// Consolidate runs a full consolidation cycle immediately.
func (p *ConsolidationPipeline) Consolidate() (ConsolidationResult, error) {
	p.mu.Lock()
	defer p.mu.Unlock()

	now := time.Now()
	stBefore := len(p.shortTerm)
	ltBefore := len(p.longTerm)

	// 1. Apply decay to all memories
	allST := make([]*MemoryRecord, 0, len(p.shortTerm))
	for _, r := range p.shortTerm {
		allST = append(allST, r)
	}
	allLT := make([]*MemoryRecord, 0, len(p.longTerm))
	for _, r := range p.longTerm {
		allLT = append(allLT, r)
	}
	p.decay.DecayAll(allST, now)
	p.decay.DecayAll(allLT, now)

	// 2. Promote eligible short-term memories to long-term
	var toPromote []string
	for id, r := range p.shortTerm {
		age := now.Sub(r.CreatedAt)
		if age >= p.cfg.MinAge && r.Importance >= p.cfg.PromotionThreshold {
			toPromote = append(toPromote, id)
		}
	}
	consolidated := 0
	for _, id := range toPromote {
		r := p.shortTerm[id]
		delete(p.shortTerm, id)
		r.Consolidated = true
		p.longTerm[id] = r
		consolidated++
	}

	// 3. Prune low-importance short-term memories
	var toPrune []string
	for id, r := range p.shortTerm {
		if p.decay.ShouldPrune(r, p.cfg.PruneThreshold) {
			toPrune = append(toPrune, id)
		}
	}
	for _, id := range toPrune {
		r := p.shortTerm[id]
		delete(p.shortTerm, id)
		p.index.Remove(*r)
	}
	pruned := len(toPrune)

	p.lastConsolidation = now
	result := ConsolidationResult{
		RunID:           uuid.New().String(),
		Timestamp:       now,
		ShortTermBefore: stBefore,
		ShortTermAfter:  len(p.shortTerm),
		LongTermBefore:  ltBefore,
		LongTermAfter:   len(p.longTerm),
		Pruned:          pruned,
		Consolidated:    consolidated,
	}
	p.history = append(p.history, result)

	slog.Info("consolidation run",
		"run_id", result.RunID[:8],
		"promoted", consolidated,
		"pruned", pruned,
		"short_term", result.ShortTermAfter,
		"long_term", result.LongTermAfter,
	)
	return result, nil
}

// ShortTermSize returns the current short-term store size.
func (p *ConsolidationPipeline) ShortTermSize() int {
	p.mu.Lock()
	defer p.mu.Unlock()
	return len(p.shortTerm)
}

// LongTermSize returns the current long-term store size.
func (p *ConsolidationPipeline) LongTermSize() int {
	p.mu.Lock()
	defer p.mu.Unlock()
	return len(p.longTerm)
}

// LastConsolidationResult returns the most recent result, or nil.
func (p *ConsolidationPipeline) LastConsolidationResult() *ConsolidationResult {
	p.mu.Lock()
	defer p.mu.Unlock()
	if len(p.history) == 0 {
		return nil
	}
	r := p.history[len(p.history)-1]
	return &r
}

// Stats returns a summary of the pipeline state.
func (p *ConsolidationPipeline) Stats() map[string]any {
	return map[string]any{
		"short_term":          p.ShortTermSize(),
		"long_term":           p.LongTermSize(),
		"consolidation_runs":  len(p.history),
		"index":               p.index.IndexStats(),
	}
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func addToSet(m map[string]map[string]struct{}, key, id string) {
	if m[key] == nil {
		m[key] = make(map[string]struct{})
	}
	m[key][id] = struct{}{}
}

func addToSet64(m map[int64]map[string]struct{}, key int64, id string) {
	if m[key] == nil {
		m[key] = make(map[string]struct{})
	}
	m[key][id] = struct{}{}
}

func removeFromSet(m map[string]map[string]struct{}, key, id string) {
	if s, ok := m[key]; ok {
		delete(s, id)
	}
}

func removeFromSet64(m map[int64]map[string]struct{}, key int64, id string) {
	if s, ok := m[key]; ok {
		delete(s, id)
	}
}

func copySet(s map[string]struct{}) map[string]struct{} {
	out := make(map[string]struct{}, len(s))
	for k := range s {
		out[k] = struct{}{}
	}
	return out
}

File: omega/core/memory_consolidation.py
"""
omega.core.memory_consolidation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Consolidation pipeline for the Omega memory system.

Components
----------
  ConsolidationPipeline : Periodically merges short-term (episodic) memories into
                          long-term (semantic) storage.
  MemoryIndex           : Fast lookup by regime tag, time range, or signal type.
  ImportanceScorer      : Ranks memories by relevance to the current regime.
  DecayManager          : Exponential decay for old memories; landmark events are
                          protected from decay.

Design notes
------------
- Pure Python stdlib only.
- All time values are UNIX timestamps (float seconds).
- Decay half-life is configurable; landmark protection is implemented as a minimum
  floor so protected memories never fall below a threshold importance.
"""

from __future__ import annotations

import logging
import math
import time
import uuid
from collections import defaultdict
from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

logger = logging.getLogger("omega.core.memory_consolidation")


# ---------------------------------------------------------------------------
# Data structures
# ---------------------------------------------------------------------------


@dataclass
class MemoryRecord:
    """A single memory entry that may live in short-term or long-term storage."""

    memory_id: str
    created_at: float  # UNIX timestamp
    last_accessed: float  # updated on read
    importance: float  # 0.0-1.0; decays or is reinforced over time
    regime_tag: str  # regime label at time of creation
    signal_type: str  # e.g. "price_move", "volatility_spike", "regime_shift"
    content: dict[str, Any]
    tags: list[str]
    is_landmark: bool = False  # if True, protected from decay below floor
    access_count: int = 0
    consolidated: bool = False  # True once moved to long-term store

    @classmethod
    def create(
        cls,
        regime_tag: str,
        signal_type: str,
        content: dict[str, Any],
        tags: list[str] | None = None,
        importance: float = 0.5,
        is_landmark: bool = False,
    ) -> MemoryRecord:
        now = time.time()
        return cls(
            memory_id=str(uuid.uuid4()),
            created_at=now,
            last_accessed=now,
            importance=importance,
            regime_tag=regime_tag,
            signal_type=signal_type,
            content=content,
            tags=tags or [],
            is_landmark=is_landmark,
        )


# ---------------------------------------------------------------------------
# MemoryIndex
# ---------------------------------------------------------------------------


class MemoryIndex:
    """
    In-memory inverted index for fast lookup by:
      - regime_tag   (e.g. "volatile", "trending")
      - signal_type  (e.g. "price_move", "regime_shift")
      - time range   (bucketed by day for O(days) range queries)

    All three lookup methods return sets of memory_ids.
    """

    _BUCKET_SIZE = 86_400  # one day in seconds

    def __init__(self) -> None:
        self._by_regime: dict[str, set[str]] = defaultdict(set)
        self._by_signal: dict[str, set[str]] = defaultdict(set)
        # time bucket → set of memory_ids
        self._by_time: dict[int, set[str]] = defaultdict(set)

    # ------------------------------------------------------------------
    # Mutation
    # ------------------------------------------------------------------

    def add(self, record: MemoryRecord) -> None:
        self._by_regime[record.regime_tag].add(record.memory_id)
        self._by_signal[record.signal_type].add(record.memory_id)
        bucket = int(record.created_at // self._BUCKET_SIZE)
        self._by_time[bucket].add(record.memory_id)

    def remove(self, record: MemoryRecord) -> None:
        self._by_regime[record.regime_tag].discard(record.memory_id)
        self._by_signal[record.signal_type].discard(record.memory_id)
        bucket = int(record.created_at // self._BUCKET_SIZE)
        self._by_time[bucket].discard(record.memory_id)

    # ------------------------------------------------------------------
    # Lookup
    # ------------------------------------------------------------------

    def by_regime(self, regime_tag: str) -> set[str]:
        return set(self._by_regime.get(regime_tag, set()))

    def by_signal(self, signal_type: str) -> set[str]:
        return set(self._by_signal.get(signal_type, set()))

    def by_time_range(self, start: float, end: float) -> set[str]:
        start_bucket = int(start // self._BUCKET_SIZE)
        end_bucket = int(end // self._BUCKET_SIZE)
        ids: set[str] = set()
        for b in range(start_bucket, end_bucket + 1):
            ids.update(self._by_time.get(b, set()))
        return ids

    def intersection(self, *id_sets: set[str]) -> set[str]:
        if not id_sets:
            return set()
        result = id_sets[0]
        for s in id_sets[1:]:
            result = result & s
        return result

    def stats(self) -> dict[str, Any]:
        return {
            "regimes": {k: len(v) for k, v in self._by_regime.items()},
            "signal_types": {k: len(v) for k, v in self._by_signal.items()},
            "time_buckets": len(self._by_time),
        }


# ---------------------------------------------------------------------------
# DecayManager
# ---------------------------------------------------------------------------


class DecayManager:
    """
    Applies exponential importance decay to MemoryRecords.

    Importance at time t:
        I(t) = max(floor, I0 x exp(-lambda x dt))

    where λ = ln(2) / half_life_seconds.

    Landmark memories have a protected floor of `landmark_floor` (default 0.3)
    so they are never pruned even after long periods of inactivity.
    """

    def __init__(
        self,
        half_life_seconds: float = 7 * 86_400,  # 7 days default
        floor: float = 0.0,
        landmark_floor: float = 0.3,
    ) -> None:
        if half_life_seconds <= 0:
            raise ValueError("half_life_seconds must be positive")
        self._decay_lambda = math.log(2) / half_life_seconds
        self._floor = floor
        self._landmark_floor = landmark_floor

    def decay(self, record: MemoryRecord, now: float | None = None) -> None:
        """Mutates record.importance in place."""
        if now is None:
            now = time.time()
        elapsed = max(0.0, now - record.created_at)
        decayed = record.importance * math.exp(-self._decay_lambda * elapsed)
        floor = self._landmark_floor if record.is_landmark else self._floor
        record.importance = max(floor, decayed)

    def decay_all(self, records: Iterable[MemoryRecord], now: float | None = None) -> None:
        if now is None:
            now = time.time()
        for r in records:
            self.decay(r, now)

    def should_prune(self, record: MemoryRecord, threshold: float = 0.05) -> bool:
        """Returns True if the record importance is below threshold and not a landmark."""
        if record.is_landmark:
            return False
        return record.importance < threshold


# ---------------------------------------------------------------------------
# ImportanceScorer
# ---------------------------------------------------------------------------


class ImportanceScorer:
    """
    Ranks MemoryRecords by relevance to the current regime.

    Scoring formula (all components in [0, 1]):
        score = w_importance x importance
              + w_recency    x recency_score
              + w_regime     x regime_match
              + w_access     x access_score

    - recency_score  : exp(-lambda x age_seconds) with a 30-day half-life
    - regime_match   : 1.0 if record.regime_tag == current_regime else 0.0
    - access_score   : tanh(access_count / 10)  — rewards frequently accessed memories
    """

    _RECENCY_HALF_LIFE = 30 * 86_400  # 30 days

    def __init__(
        self,
        w_importance: float = 0.4,
        w_recency: float = 0.3,
        w_regime: float = 0.2,
        w_access: float = 0.1,
    ) -> None:
        total = w_importance + w_recency + w_regime + w_access
        if abs(total - 1.0) > 1e-6:
            raise ValueError(f"Weights must sum to 1.0, got {total}")
        self._wi = w_importance
        self._wr = w_recency
        self._wg = w_regime
        self._wa = w_access
        self._recency_lambda = math.log(2) / self._RECENCY_HALF_LIFE

    def score(
        self,
        record: MemoryRecord,
        current_regime: str,
        now: float | None = None,
    ) -> float:
        if now is None:
            now = time.time()

        age = max(0.0, now - record.created_at)
        recency = math.exp(-self._recency_lambda * age)
        regime_match = 1.0 if record.regime_tag == current_regime else 0.0
        access = math.tanh(record.access_count / 10.0)

        return (
            self._wi * record.importance
            + self._wr * recency
            + self._wg * regime_match
            + self._wa * access
        )

    def rank(
        self,
        records: Iterable[MemoryRecord],
        current_regime: str,
        now: float | None = None,
        top_k: int | None = None,
    ) -> list[tuple[float, MemoryRecord]]:
        """Returns (score, record) pairs sorted descending by score."""
        if now is None:
            now = time.time()
        scored = [(self.score(r, current_regime, now), r) for r in records]
        scored.sort(key=lambda x: x[0], reverse=True)
        if top_k is not None:
            scored = scored[:top_k]
        return scored


# ---------------------------------------------------------------------------
# ConsolidationPipeline
# ---------------------------------------------------------------------------


@dataclass
class ConsolidationResult:
    """Summary of a single consolidation run."""

    run_id: str
    timestamp: float
    short_term_before: int
    short_term_after: int
    long_term_before: int
    long_term_after: int
    pruned: int
    consolidated: int


class ConsolidationPipeline:
    """
    Periodically merges short-term memories into long-term storage.

    Short-term → long-term promotion criteria:
      1. Memory is older than `min_age_seconds`.
      2. Memory importance is above `promotion_threshold`.

    After promotion, memories in short-term that are below `prune_threshold`
    importance are pruned.

    The pipeline also manages a MemoryIndex, a DecayManager, and an
    ImportanceScorer so callers only need to interact with one object.
    """

    def __init__(
        self,
        min_age_seconds: float = 3_600,  # 1 hour before eligible for consolidation
        promotion_threshold: float = 0.3,
        prune_threshold: float = 0.05,
        consolidation_interval: float = 900,  # run every 15 minutes
        decay_half_life: float = 7 * 86_400,
        landmark_floor: float = 0.3,
    ) -> None:
        self._min_age = min_age_seconds
        self._promotion_threshold = promotion_threshold
        self._prune_threshold = prune_threshold
        self._interval = consolidation_interval

        self.index = MemoryIndex()
        self.decay_manager = DecayManager(
            half_life_seconds=decay_half_life,
            landmark_floor=landmark_floor,
        )
        self.scorer = ImportanceScorer()

        # short_term: unsettled / recently created
        self._short_term: dict[str, MemoryRecord] = {}
        # long_term: consolidated / stable
        self._long_term: dict[str, MemoryRecord] = {}

        self._last_consolidation: float = time.time()
        self._history: list[ConsolidationResult] = []

    # ------------------------------------------------------------------
    # Ingestion
    # ------------------------------------------------------------------

    def add(self, record: MemoryRecord) -> None:
        """Add a new memory to short-term storage and index it."""
        self._short_term[record.memory_id] = record
        self.index.add(record)
        logger.debug(
            "Added memory %s (regime=%s signal=%s importance=%.2f)",
            record.memory_id[:8],
            record.regime_tag,
            record.signal_type,
            record.importance,
        )

    # ------------------------------------------------------------------
    # Retrieval
    # ------------------------------------------------------------------

    def get(self, memory_id: str) -> MemoryRecord | None:
        record = self._short_term.get(memory_id) or self._long_term.get(memory_id)
        if record is not None:
            record.last_accessed = time.time()
            record.access_count += 1
        return record

    def query(
        self,
        regime_tag: str | None = None,
        signal_type: str | None = None,
        start_time: float | None = None,
        end_time: float | None = None,
        current_regime: str = "",
        top_k: int = 20,
    ) -> list[tuple[float, MemoryRecord]]:
        """Multi-criteria query returning top-k records by importance score."""
        candidate_ids: set[str] | None = None

        def _intersect(ids: set[str]) -> None:
            nonlocal candidate_ids
            if candidate_ids is None:
                candidate_ids = ids
            else:
                candidate_ids &= ids

        if regime_tag is not None:
            _intersect(self.index.by_regime(regime_tag))
        if signal_type is not None:
            _intersect(self.index.by_signal(signal_type))
        if start_time is not None and end_time is not None:
            _intersect(self.index.by_time_range(start_time, end_time))

        if candidate_ids is None:
            all_records: list[MemoryRecord] = list(self._short_term.values()) + list(
                self._long_term.values()
            )
        else:
            all_records = [r for mid in candidate_ids if (r := self.get(mid)) is not None]

        return self.scorer.rank(all_records, current_regime, top_k=top_k)

    # ------------------------------------------------------------------
    # Consolidation
    # ------------------------------------------------------------------

    def maybe_consolidate(self, now: float | None = None) -> ConsolidationResult | None:
        """Run consolidation if the interval has elapsed; otherwise no-op."""
        if now is None:
            now = time.time()
        if now - self._last_consolidation < self._interval:
            return None
        return self.consolidate(now)

    def consolidate(self, now: float | None = None) -> ConsolidationResult:
        """Run a full consolidation cycle immediately."""
        if now is None:
            now = time.time()

        st_before = len(self._short_term)
        lt_before = len(self._long_term)

        # 1. Apply decay to all memories
        self.decay_manager.decay_all(self._short_term.values(), now)
        self.decay_manager.decay_all(self._long_term.values(), now)

        # 2. Promote eligible short-term memories to long-term
        consolidated = 0
        to_promote: list[str] = []
        for mid, record in self._short_term.items():
            age = now - record.created_at
            if age >= self._min_age and record.importance >= self._promotion_threshold:
                to_promote.append(mid)

        for mid in to_promote:
            record = self._short_term.pop(mid)
            record.consolidated = True
            self._long_term[mid] = record
            consolidated += 1

        # 3. Prune low-importance short-term memories
        pruned = 0
        to_prune: list[str] = []
        for mid, record in self._short_term.items():
            if self.decay_manager.should_prune(record, self._prune_threshold):
                to_prune.append(mid)

        for mid in to_prune:
            record = self._short_term.pop(mid)
            self.index.remove(record)
            pruned += 1

        self._last_consolidation = now
        result = ConsolidationResult(
            run_id=str(uuid.uuid4()),
            timestamp=now,
            short_term_before=st_before,
            short_term_after=len(self._short_term),
            long_term_before=lt_before,
            long_term_after=len(self._long_term),
            pruned=pruned,
            consolidated=consolidated,
        )
        self._history.append(result)
        logger.info(
            "Consolidation run %s: promoted=%d pruned=%d ST=%d LT=%d",
            result.run_id[:8],
            consolidated,
            pruned,
            result.short_term_after,
            result.long_term_after,
        )
        return result

    # ------------------------------------------------------------------
    # Introspection
    # ------------------------------------------------------------------

    @property
    def short_term_size(self) -> int:
        return len(self._short_term)

    @property
    def long_term_size(self) -> int:
        return len(self._long_term)

    def last_consolidation_result(self) -> ConsolidationResult | None:
        return self._history[-1] if self._history else None

    def stats(self) -> dict[str, Any]:
        return {
            "short_term": self.short_term_size,
            "long_term": self.long_term_size,
            "consolidation_runs": len(self._history),
            "index": self.index.stats(),
        }
